### DISCUSSION TITLE
Question about `redirect` and `redirect-rule` filters

### DISCUSSION THREAD

**Discussion by partingscientist** [uBlockOrigin/uAssets#20581] [Q&A]
I was investigating `https://github.com/easylist/easylist/issues/16421` earlier. My question is, can one do better than `@@||pagead2.googlesyndication.com/pagead/show_ads.js$script,domain=ageofconsent.net`? I thought that perhaps `redirect` or `redirect-rule` can be used, but I'm not very familiar with them. Thank you.

**Comment by MasterKia** [comment]
This is already in `uBlock filters - Annoyances`:
```adb
||pagead2.googlesyndication.com/pagead/show_ads.js$script,redirect-rule=noopjs,from=ageofconsent.net
```

**Reply by partingscientist** [reply]
I'm pretty sure I tested something similar to this earlier, yet it doesn't work. The one I tested earlier must've had a mistake in it because this one works. Thank you!

**Comment by stephenhawk8054** [comment]
`redirect` blocks and redirects the resource
`redirect-rule` only redirects the blocked resource

Currently there's already a `redirect-rule` in annoyance list as shown above, but if there's an exception filter (by EL or something), you can override it by `redirect` + `important`

```adb
||pagead2.googlesyndication.com/pagead/show_ads.js$script,domain=ageofconsent.net,redirect=noopjs,important
```

**Reply by partingscientist** [reply]
Thank you!

**Reply by krystian3w** [reply]
EL anti-ads exception should be covered in `ublock*.txt`

**Comment by partingscientist** [comment]
I'm still confused with these two modifiers. I would expect the following filter
```adblock
||example.com^$redirect=noophtml
```
would cause `example.com` to be replaced with an empty HTML page upon visiting, which is not the case. Is this not the correct use case for the modifiers? Is my use case even possible to be achieved with filters?

**Reply by stephenhawk8054** [reply]
`redirect` does not support `doc`

**Reply by partingscientist** [reply]
Well that's unfortunate. I wasn't able to find any information about that in the wiki also. Thank you regardless.

**Comment by partingscientist** [comment]
Does uBO support hostname redirection similar to `https://adguard.com/kb/general/ad-filtering/create-own-filters/#urltransform-modifier:~:text=Changing%20the%20origin`?

**Reply by gorhill** [reply]
`uritransform` is a filter option requiring trust, so you can rewrite any portion of the URL to transform. The documentation in the wiki needs to be brought up to date, it's no longer accurate.

---

For example, this works:

    ||example.com^$uritransform=/\.com/.org/

**Reply by partingscientist** [reply]
I thought `uritransform` can only change URL's path. That's good to know.

**Reply by gorhill** [reply]
This was changed at some point, can't remember exactly, would have to dig with `git blame`. If I remember correctly, the change was because DNR rules do not have origin restrictions about transforming a URL into another (except for `script` resources), so there was no point enforcing one in uBO given the filter already requires trust.

**Reply by gorhill** [reply]
Here: <https://github.com/gorhill/uBlock/commit/aaf35d9d7106800ed2dfdded3b40d7afa404e464>

**Reply by stephenhawk8054** [reply]
Yeah, we needed `uritransform` [to deal with ads on Amazon prime for uBOL](https://github.com/uBlockOrigin/uBlock-discussions/discussions/913#discussioncomment-14106157) (internal link). But since DNR syntax could replace the domains as well, `uritransform` was adjusted for compatibility.
